be/src/exec/orc_scanner.cpp

```diff
@@ -0,0 +1,298 @@
+    }
+
+    int64_t bytes_read = 0;
+    int64_t reads = 0;
+    while (bytes_read < length) {
+        Status result = _file->readat(offset, length - bytes_read, &reads, buf);
+        if (!result.ok()) {
+            throw orc::ParseError("Bad read of " + _filename);
+        }
+        if (reads == 0) {
+            break;
+        }
+        bytes_read += reads;// total read bytes
+        offset += reads;
+        buf = (char *) buf + reads;
+    }
+    if (length != bytes_read) {
+        throw orc::ParseError("Short read of " + _filename
+                              + ". expected :" + std::to_string(length) + ", actual : " + std::to_string(bytes_read));
+    }
+}
+
+const std::string& ORCFileStream::getName() const {
+    return _filename;
+}
+
+ORCScanner::ORCScanner(RuntimeState *state,
+        RuntimeProfile *profile,
+        const TBrokerScanRangeParams &params,
+        const std::vector<TBrokerRangeDesc> &ranges,
+        const std::vector<TNetworkAddress> &broker_addresses,
+        ScannerCounter *counter) : BaseScanner(state, profile, params, counter),
+            _ranges(ranges),
+            _broker_addresses(broker_addresses),
+            // _splittable(params.splittable),
+            _next_range(0),
+            _cur_file_eof(true),
+            _scanner_eof(false),
+            _total_groups(0),
+            _current_group(0),
+            _rows_of_group(0),
+            _current_line_of_group(0) {}
+
+ORCScanner::~ORCScanner() {
+    close();
+}
+
+Status ORCScanner::open() {
+    return BaseScanner::open();
+}
+
+Status ORCScanner::get_next(Tuple *tuple, MemPool *tuple_pool, bool *eof) {
+    try {
+        SCOPED_TIMER(_read_timer);
+        // Get one line
+        while (!_scanner_eof) {
+            if (_cur_file_eof) {
+                RETURN_IF_ERROR(open_next_reader());
+                if (_scanner_eof) {
+                    *eof = true;
+                    return Status::OK();
+                } else {
+                    _cur_file_eof = false;
+                }
+            }
+            if (_current_line_of_group >= _rows_of_group) { // read next stripe
+                if (_current_group >= _total_groups) {
+                    _cur_file_eof = true;
+                    continue;
+                }
+                _rows_of_group = _reader->getStripe(_current_group)->getNumberOfRows();
+                _batch = _row_reader->createRowBatch(_rows_of_group);
+                _row_reader->next(*_batch.get());
+
+                _current_line_of_group = 0;
+                ++_current_group;
+            }
+
+            const std::vector<orc::ColumnVectorBatch *>& batch_vec = ((orc::StructVectorBatch *) _batch.get())->fields;
+            for (int column_ipos = 0; column_ipos < _num_of_columns_from_file; ++column_ipos) {
+                auto slot_desc = _src_slot_descs[column_ipos];
+                int position_in_orc = _column_name_map_orc_index.find(slot_desc->col_name())->second;
```

Comment:
we do this map find for each fields of each row, I think it is bad for performance.
Because the column handle serial is the same for every row, we can get the positition vector when preparing.

Reply:
Done